Const Block Warm-up

iten-alg opened the pull request:
Addresses the bytecblock portions of issue 4034 where the assembler has bad interactions with the byte pseudo-ops. 

Remain to be addressed in later updates tonight: intcblock side of things, disassembler, and what to do if exactly one const block manually created but unclear whether block will be made before or after pseudo-ops

iten-alg committed:
bytecblock stuff

iten-alg committed:
Merge branch 'master' into warm_up

iten-alg committed:
Added intc side of things, still need test cases

iten-alg committed:
Added int/intcblock test cases

jannotti commented on data/transactions/logic/assembler_test.go:
```diff
@@ -2233,6 +2233,28 @@ func TestAssembleConstants(t *testing.T) {
+	ops := testProg(t, "intcblock 0x01 0x02; int 0x04", AssemblerMaxVersion)
+	expected := []byte{byte(AssemblerMaxVersion), 0x20, 0x02, 0x01, 0x02, 0x81, 0x04}
+	require.Equal(t, expected, ops.Program)
+	// If the int is already in the block, we mind as well use the block
```
mind -> might

jannotti commented on data/transactions/logic/assembler_test.go:
```diff
@@ -2233,6 +2233,28 @@ func TestAssembleConstants(t *testing.T) {
+	ops = testProg(t, "intcblock 0x01 0x02; int 0x02", AssemblerMaxVersion)
+	expected = []byte{byte(AssemblerMaxVersion), 0x20, 0x02, 0x01, 0x02, 0x23}
+	require.Equal(t, expected, ops.Program)
+	// Previous int pseudo-ops should be pushints
```
We probably have a tough time really knowing what's "previous" here, right?  In this example it's correct, but with branching, you could have some code that appears before other code, but runs later.

iten-alg replied:
I agree, but it's also pretty nonsensical to only have a single intcblock but then also not put it at the beginning (except for some potential dynamic cost saving I suppose). I'm happy to change it to just if you manually put in a const block, the pseudo-ops all go to pushint/pushbytes, but this means that people who want to manually set up const blocks now also have to manually index them.

Another option could be to require the const block setups to be the first ops or else the pseudo-ops just go to pushint/pushbytes

iten-alg committed:
small change

iten-alg committed:
small change

iten-alg committed:
nearly everything done
todo: add test cases for byte constant disassembly, handle terminating const blocks

iten-alg committed:
bug fix and added byte constant disassembly test cases

iten-alg committed:
All functionality and test cases should be implemented now

iten-alg committed:
Clean up
